service/storage_service.cc

```diff
@@ -315,6 +317,9 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
         app_states.emplace(gms::application_state::TOKENS, value_factory.tokens(_bootstrap_tokens));
         app_states.emplace(gms::application_state::CDC_STREAMS_TIMESTAMP, value_factory.cdc_streams_timestamp(_cdc_streams_ts));
         app_states.emplace(gms::application_state::STATUS, value_factory.hibernate(true));
+        auto replace_address = db().local().get_replace_address();
+        replacing_a_node_with_same_ip = replace_address && *replace_address == get_broadcast_address();
+        replacing_a_node_with_diff_ip = replace_address && *replace_address != get_broadcast_address();
```

Comment:
Hi @asias.

The explanation you give in the changelog is not enough for a person who doesn't understand this code that well (me) to understand why we really need this differentiation.

It feels weird to me that we have to make this distinction. What would be the effect of us not differentiating between nodes with the same IP and different IP?

Reply:
The key point is we want the replacing node has the same view of the token ring as the result of the cluster. In case the replacing node is started with a new ip, the result of the cluster will put the new ip into the token_metadata (ip token map) after the replacing is done. In case the replacing node is started with the same IP ad the replaced node, the rest of the cluster as the old ip in token_metadata, so should the replacing node.